### PR TITLE
Feature - get issuer

### PR DESCRIPTION
**Pull request by brkagithub**
Added getStateIssuer, getLatestStateIssuer and getAssertionIds, also additional helper functions in blockchain service and validation service. Update was not working correctly on my local node setup for some reason so should test with update before merging, I'll organize this with Nikola before we merge.

### CODE REVIEW

File: managers/asset-operations-manager.js [hunk 1]
```diff
@@ -790,32 +790,89 @@ class AssetOperationsManager {
...
 
-        const issuer = await this.blockchainService.getAssertionIssuer(tokenId, firstAssertionId, assertionIndex, blockchain);
+        const issuer = await this.blockchainService.getAssertionIssuer(
+            tokenId,
+            assertionId,
+            stateIndex,
+            blockchain,
+        );
+        return {
+            UAL,
```
Comment (u-hubar) [inline managers/asset-operations-manager.js:819]: It makes to return state as well imo, just so user has a possibility to check that it's the state he wanted

File: managers/asset-operations-manager.js [hunk 2]
```diff
@@ -790,32 +790,89 @@ class AssetOperationsManager {
...
+        let indexOfLastAssertion = assertionIds.length - 1;
+
+        const issuer = await this.blockchainService.getAssertionIssuer(
+            tokenId,
+            assertionIds[indexOfLastAssertion],
+            indexOfLastAssertion,
+            blockchain,
+        );
         return {
             UAL,
```
Comment (u-hubar) [inline managers/asset-operations-manager.js:849]: Let's return state here as well

File: managers/asset-operations-manager.js [hunk 3]
```diff
@@ -790,32 +790,89 @@ class AssetOperationsManager {
...
+     * @param {Object} [options={}] - Optional parameters for blockchain service.
+     * @returns {Object} An object containing the UAL, issuer and operation status.
+     */
+    async getLatestStateIssuer(UAL, options = {}) {
+        const blockchain = this.inputService.getBlockchain(options);
+        this.validationService.validateAssetGetLatestStateIssuer(UAL, blockchain);
+
+        const { tokenId } = resolveUAL(UAL);
+
+        const assertionIds = await this.blockchainService.getAssertionIds(tokenId, blockchain);
```
Comment (u-hubar) [inline managers/asset-operations-manager.js:838]:
```suggestion
        const states = await this.blockchainService.getAssertionIds(tokenId, blockchain);
```

File: managers/asset-operations-manager.js [hunk 4]
```diff
@@ -790,32 +790,89 @@ class AssetOperationsManager {
...
+
+        let indexOfLastAssertion = assertionIds.length - 1;
```
Comment (u-hubar) [inline managers/asset-operations-manager.js:840]:
```suggestion
        const latestStateIndex = assertionIds.length - 1;
        const latestState = states[latestStateIndex];
```

File: managers/asset-operations-manager.js (hunk 2 above)
Comment (u-hubar) [inline managers/asset-operations-manager.js:849]:
```suggestion
            UAL,
            latestState,
```

File: managers/asset-operations-manager.js [hunk 5]
```diff
@@ -790,32 +790,89 @@ class AssetOperationsManager {
...
+     */
+    async getStates(UAL, options = {}) {
+        const blockchain = this.inputService.getBlockchain(options);
+        this.validationService.validateAssetGetStates(UAL, blockchain);
+
+        const { tokenId } = resolveUAL(UAL);
+
+        const states = await this.blockchainService.getAssertionIds(tokenId, blockchain);
+
+        return {
```
Comment (u-hubar) [inline managers/asset-operations-manager.js:870]:
```suggestion
        return {
            UAL,
```

File: managers/asset-operations-manager.js [hunk 6]
```diff
@@ -790,32 +790,89 @@ class AssetOperationsManager {
...
+    async getStateIssuer(UAL, stateIndex, options = {}) {
         const blockchain = this.inputService.getBlockchain(options);
-        this.validationService.validateAssetGetIssuer(UAL, blockchain);
+        this.validationService.validateAssetGetStateIssuer(UAL, blockchain);
 
         const { tokenId } = resolveUAL(UAL);
-        const assertionIndex = 0;
-        const firstAssertionId = await this.blockchainService.getAssertionIdByIndex(
+
+        const assertionId = await this.blockchainService.getAssertionIdByIndex(
```
Comment (u-hubar) [inline managers/asset-operations-manager.js:806]:
```suggestion
        const state = await this.blockchainService.getAssertionIdByIndex(
```

File: managers/asset-operations-manager.js (hunk 1 above)
Comment (u-hubar) [inline managers/asset-operations-manager.js:819]:
```suggestion
            UAL,
            state,
```

File: managers/asset-operations-manager.js [hunk 7]
```diff
@@ -790,32 +790,89 @@ class AssetOperationsManager {
...
             tokenId,
-            assertionIndex,
+            stateIndex,
             blockchain,
         );
 
-        const issuer = await this.blockchainService.getAssertionIssuer(tokenId, firstAssertionId, assertionIndex, blockchain);
+        const issuer = await this.blockchainService.getAssertionIssuer(
+            tokenId,
+            assertionId,
```
Comment (u-hubar) [inline managers/asset-operations-manager.js:814]:
```suggestion
            state,
```

File: managers/asset-operations-manager.js [hunk 8]
```diff
@@ -790,32 +790,89 @@ class AssetOperationsManager {
...
+
+        const { tokenId } = resolveUAL(UAL);
+
+        const assertionIds = await this.blockchainService.getAssertionIds(tokenId, blockchain);
+
+        let indexOfLastAssertion = assertionIds.length - 1;
+
+        const issuer = await this.blockchainService.getAssertionIssuer(
+            tokenId,
+            assertionIds[indexOfLastAssertion],
```
Comment (u-hubar) [inline managers/asset-operations-manager.js:844]:
```suggestion
            latestState,
```

File: services/validation-service.js [hunk 9]
```diff
@@ -148,7 +148,17 @@ class ValidationService {
         this.validateBlockchain(blockchain);
     }
 
-    validateAssetGetIssuer(UAL, blockchain) {
+    validateAssetGetStateIssuer(UAL, blockchain) {
```
Comment (u-hubar) [inline services/validation-service.js:151]: I guess we should validate stateIndex here as well?

**Review by u-hubar — CHANGES_REQUESTED** [review]
Overall looks good, just some small suggestions how to improve it further

**Review by NZT48 — APPROVED** [review]

**Review by djordjekovac — APPROVED** [review]

**Pull request merged**